rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,418 @@
+
+This join method is explicitly designed to be used with Terraform and IaC
+workflows. Today, it's possible to generate one or more secret values, compute
+an expiration time, provision a token for each secret, and spawn many VM
+instances with that token passed along via user data. While a bit verbose, this
+initial deployment workflow is mostly sane. For an example, refer to [this
+documentation
+snapshot](https://github.com/gravitational/teleport/blob/cb0a69d09550e45c2c327ab7dcc6a023e3bb162a/docs/pages/reference/terraform-provider/resources/bot.mdx#example-usage)
+with a working example.
+
+However, the critical issue in this workflow is in maintenance: if any one of
+these bots expires, there is no reasonable method to restore it short of
+manually issuing a new join token (`tctl bots instance add foo`), connecting to
+the node, and manually copying the new token into the bot config.
+
+The new `challenge` method improves this situation in two primary ways: one
+fewer resource needs to be generated (the secret value), and maintenance can be
+performed simply by adjusting values in the resource. If a bot fails, its rejoin
+counter can be incremented easily.
+
+As an example, we can consider provisioning several bots. We'll need to account
+for future overrides so we can fix a single bot in the future if needed:
+
+``` hcl
+locals {
+  nodes = toset(["foo", "bar", "baz"])
+}
+
+resource "teleport_bot" "example" {
+  name = "example"
+  roles = ["access"]
+}
+
+variable "bot_rejoin_overrides" {
+  type = map(number)
+  default = {
+    foo = 5
+  }
+}
+
+resource "teleport_provision_token" "example" {
+  for_each = local.nodes
+
+  version = "v2"
+  metadata = {
+    name = "example-${each.key}"
+  }
+  spec = {
+    roles = ["Bot"]
+    bot_name = teleport_bot.example.name
+    join_method = "challenge"
+
+    challenge = {
+      rejoining = {
+        # look up node-specific count in the rejoin overrides map, default to 2
+        total_rejoins = lookup(var.bot_rejoin_overrides, each.key, 2)
+      }
+    }
+  }
+}
+
+resource "aws_instance" "example" {
+  for_each = teleport_provision_token.example
+
+  ami           = "ami-12345678"
+  instance_type = "t2.micro"
+
+  user_data = templatefile("${path.module}/user-data.tpl", {
+    initial_join_token = each.value.status.challenge.initial_join_secret
+  })
+}
+```
+
+In this example, if node `bar` uses its 2 renewals, we can add a new entry for
+it in `bot_rejoin_overrides` and its `ProvisionToken` will be updated to allow
+additional renewals.
+
+#### Challenge Ceremony
+
+The challenge ceremony will take inspiration from several existing join methods.
+We can create an interactive challenge similar to TPM joining and present a
+challenge containing a nonce. To avoid completely implementing our own
+authentication ceremony, clients can use `go-jose` to marshal and sign a JWT
+which can then be verified easily on the server.
+
+TODO: This needs significant further elaboration and feedback.
```

Comment:
Intentionally brief at the moment, I think I'd like to consult with some experts before committing to any crypto implementation. Very much open to ideas and feedback on this!

Reply:
I think JWS/JWT seems to be a sensible choice here. 

How will this flow work when the bot is providing its public key on initial join? Will it still need to perform an initial signature?

Reply:
I think so, it seems like it'd be tidier if the secret-for-pubkey exchange was effectively an independent step and then it did a full regular rejoin process.

That said, keeping them fully separate may create unnecessary complexity elsewhere, since we'd need to have yet another public RPC. I'll try to explore both options and see which is simpler and easier to keep secure.